traitsui/examples/demo/Advanced/Dynamic_range_trait_and_editor.py

```diff
@@ -135,10 +135,14 @@ def _get_min_temperature(self):
                 min(int(60.00 / self.fuel_cost), 15))
 
     # Event handlers:
-    @on_trait_change('guests[]')
-    def _guests_modified(self, removed, added):
-        for guest in added:
-            guest.hotel = self
+    @observe('guests.items')
+    def _guests_modified(self, event):
+        if isinstance(event.object, Hotel):
+            for guest in event.new:
+                guest.hotel = self
+        else:
+            for guest in event.added:
+                guest.hotel = self
```

Comment:
This is similar to the commit immediately prior.  It is unclear me how we want to handle cases where we were listening to `container[]` and now observe `container.items`, but the handler previously expeced to have `added` and `removed`.  Changes to `container` will fire a `TraitChangeEvent` which has `object`, `name`, `old`, `new`, whereas changes to an item in the container will have a `TraitListEvent` or the like which allows us to access `added`/`removed`.  Also this example currently is broken with the following error when you click "add guest":
```
ERROR:traits:Exception occurred in traits notification handler for object: <__main__.Hotel object at 0x7f8a298bb570>, trait: guests_items, old value: <undefined>, new value: TraitListEvent(index=4, removed=[], added=[<__main__.Guest object at 0x7f8a29a650f8>])
Traceback (most recent call last):
  File "/Users/aayres/.edm/envs/traitsui-test-3.6-pyqt5/lib/python3.6/site-packages/traits/trait_notifiers.py", line 524, in _dispatch_change_event
    self.dispatch(handler, *args)
  File "/Users/aayres/.edm/envs/traitsui-test-3.6-pyqt5/lib/python3.6/site-packages/traits/trait_notifiers.py", line 619, in dispatch
    handler(*args)
  File "/Users/aayres/.edm/envs/traitsui-test-3.6-pyqt5/lib/python3.6/site-packages/traitsui/qt4/list_editor.py", line 559, in update_editor_item
    ui, view_object, monitoring = self._create_page(object)
  File "/Users/aayres/.edm/envs/traitsui-test-3.6-pyqt5/lib/python3.6/site-packages/traitsui/qt4/list_editor.py", line 657, in _create_page
    parent=self.control, view=factory.view, kind=factory.ui_kind
  File "/Users/aayres/.edm/envs/traitsui-test-3.6-pyqt5/lib/python3.6/site-packages/traits/has_traits.py", line 1744, in edit_traits
    args,
  File "/Users/aayres/.edm/envs/traitsui-test-3.6-pyqt5/lib/python3.6/site-packages/traitsui/view.py", line 455, in ui
    ui.ui(parent, kind)
  File "/Users/aayres/.edm/envs/traitsui-test-3.6-pyqt5/lib/python3.6/site-packages/traitsui/ui.py", line 237, in ui
    self.rebuild(self, parent)
  File "/Users/aayres/.edm/envs/traitsui-test-3.6-pyqt5/lib/python3.6/site-packages/traitsui/qt4/toolkit.py", line 157, in ui_subpanel
    ui_panel.ui_subpanel(ui, parent)
  File "/Users/aayres/.edm/envs/traitsui-test-3.6-pyqt5/lib/python3.6/site-packages/traitsui/qt4/ui_panel.py", line 72, in ui_subpanel
    _ui_panel_for(ui, parent, True)
  File "/Users/aayres/.edm/envs/traitsui-test-3.6-pyqt5/lib/python3.6/site-packages/traitsui/qt4/ui_panel.py", line 78, in _ui_panel_for
    ui.control = control = _Panel(ui, parent, is_subpanel).control
  File "/Users/aayres/.edm/envs/traitsui-test-3.6-pyqt5/lib/python3.6/site-packages/traitsui/qt4/ui_panel.py", line 140, in __init__
    self.control = panel(ui)
  File "/Users/aayres/.edm/envs/traitsui-test-3.6-pyqt5/lib/python3.6/site-packages/traitsui/qt4/ui_panel.py", line 269, in panel
    panel = _GroupPanel(content[0], ui).control
  File "/Users/aayres/.edm/envs/traitsui-test-3.6-pyqt5/lib/python3.6/site-packages/traitsui/qt4/ui_panel.py", line 627, in __init__
    layout = self._add_items(content, inner)
  File "/Users/aayres/.edm/envs/traitsui-test-3.6-pyqt5/lib/python3.6/site-packages/traitsui/qt4/ui_panel.py", line 893, in _add_items
    ui, object, name, item.tooltip, None
  File "/Users/aayres/.edm/envs/traitsui-test-3.6-pyqt5/lib/python3.6/site-packages/traitsui/editors/range_editor.py", line 283, in simple_editor
    ui
  File "/Users/aayres/.edm/envs/traitsui-test-3.6-pyqt5/lib/python3.6/site-packages/traitsui/editors/range_editor.py", line 193, in _get_low_high
    low = self.named_value(self.low_name, ui)
  File "/Users/aayres/.edm/envs/traitsui-test-3.6-pyqt5/lib/python3.6/site-packages/traitsui/editor_factory.py", line 108, in named_value
    value = getattr(value, name)
AttributeError: 'NoneType' object has no attribute 'min_temperature'
```
This seems to originate for the `_add_guest_changed` method below.


We should not merge this PR until this is fixed

Reply:
I wonder if it's better to split the listener into two separate methods - one for `guests` and another for `guests:items` - this way, we can handle things cleanly.

Reply:
I think this question should also be discussed on the internal slack channel - so that we can settle on a consensus for what the right thing to do is. IMO, the best thing is to separate the listeners.

Reply:
Will post a question shortly after I go through the other suggested changes

Reply:
I think this specific piece of code can stay as is. We don't need to separate it into two methods - which listen to `guests` vs `guests:items`. But, i think it's definitely useful to add a comment differentiating the two branches - it took me a little time to understand which branch handled what.

Reply:
I will add a comment.  I am unsure exactly where the failure is coming from.  It seems to originate in `_add_guest_changed` not the `_guests_modified` method that we changed here.  However, this error is not observed (at least loudly) without these changes.  It is possible something had been broken perviously but on trait change didn't make a fuss about it